enhancements/windows-containers/windows-machine-config-operator-upgrades.md

```diff
@@ -0,0 +1,195 @@
+
+### Goals
+
+As part of this enhancement we plan to do the following:
+* Enable upgrading WMCO with as little disruption as possible.
+* Enable upgrading on-node Kubernetes components through newer versions of the
+  WMCO, including major version upgrades across OpenShift cluster upgrades, and
+  bug and CVE fixes during the current version.
+
+### Non-Goals
+
+In this enhancement we do not plan to support Windows operating system updates.
+
+## Proposal
+
+To allow for the WMCO to be upgraded in a safe and non-disruptive way, it is
+proposed that Windows VMs configured by previous versions of the WMCO are
+indirectly drained and terminated in batches by user created MachineHealthChecks,
+triggered by WMCO stopping the associated node's kubelet. The upgraded WMCO
+instance will then be able to configure VMs that will be created in place of
+the terminated ones.
+
+In order to allow for this to occur with minimal service disruption, it is
+recommended that users have at least two Windows nodes within the cluster.
+
+All Windows Kubernetes component updates will be tied to WMCO releases.
+
+### Justification
+
+The reason we need to cause the VMs to be fully terminated, instead of updating
+them, is that changes made to the VM done in previous versions of the WMCO may
+cause issues with the current version. If we chose to do an in place update,
+we would need to actively undo changes made to the node that may cause issues
+with the current WMCO version. The logic to keep track and undo these changes
+adds an additional layer of complexity that also adds the benefit of a quicker
+upgrade time. The risk of introducing additional issues is not worth it at this
+time, however it should be explored as the product matures.
+
+The reason we will not handle Windows operating system updates is that the
+cluster administrator provides the Windows image to create the VMs with, and
+it stays their responsiblity to provide an updated image.
+
+It is left to the user to create the Windows MachineHealthChecks, as
+WMCO's job is solely to configure Windows VMs into nodes. It is the user's
+responsibility to create the MachineSet which in turn creates the Windows VMs.
+Because of that, the responsibility falls onto the user to set up
+MachineHealthChecks, as WMCO does not create the Machines, and should not
+decide when they should be destroyed.
+
+### Design Details
+
+WMCO is published to OperatorHub as a Red Hat operator. When a new WMCO version
+is released, and the new version is compatible with the current cluster
+version, an update can occur. While we cannot explicity specify a required OCP
+version for the WMCO upgrade, each new OCP version upgrade should include a
+Kubernetes version upgrade. This allows us to use the [minKubeVersion](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md#operator-metadata)
```

Comment:
The downside of  using `minKubeVersion` is that older operators will be able to run on new clusters. How are you going to address that?

Reply:
Just addressed in risks section

Reply:
I also feel I address this in the version skew section to an extent